src/run/handlers/use-cache-handler.ts

```diff
@@ -0,0 +1,270 @@
+const cacheHandlersSymbol = Symbol.for('@next/cache-handlers')
+const extendedGlobalThis = globalThis as typeof globalThis & {
+  // Used by Next Runtime to ensure we have single instance of
+  //  - LRUCache
+  //  - pending sets
+  // even if this module gets copied multiple times
+  [LRU_CACHE_GLOBAL_KEY]?: CacheHandleLRUCache
+  [PENDING_SETS_GLOBAL_KEY]?: PendingSets
+
+  // Used by Next.js to provide implementation of cache handlers
+  [cacheHandlersSymbol]?: {
+    RemoteCache?: CacheHandler
+    DefaultCache?: CacheHandler
+  }
+}
+
+function getLRUCache(): CacheHandleLRUCache {
+  if (extendedGlobalThis[LRU_CACHE_GLOBAL_KEY]) {
+    return extendedGlobalThis[LRU_CACHE_GLOBAL_KEY]
+  }
+
+  const lruCache = new LRUCache<string, PrivateCacheEntry>({
+    max: 1000,
+    maxSize: 50 * 1024 * 1024, // same as hardcoded default in Next.js
+    sizeCalculation: (value) => value.size,
+  })
+
+  extendedGlobalThis[LRU_CACHE_GLOBAL_KEY] = lruCache
+
+  return lruCache
+}
+
+function getPendingSets(): PendingSets {
+  if (extendedGlobalThis[PENDING_SETS_GLOBAL_KEY]) {
+    return extendedGlobalThis[PENDING_SETS_GLOBAL_KEY]
+  }
+
+  const pendingSets = new Map()
+  extendedGlobalThis[PENDING_SETS_GLOBAL_KEY] = pendingSets
+  return pendingSets
+}
+
+// eslint-disable-next-line @typescript-eslint/no-empty-function
+const tmpResolvePendingBeforeCreatingAPromise = () => {}
+
+export const NetlifyDefaultUseCacheHandler = {
+  get(cacheKey: string): ReturnType<CacheHandler['get']> {
+    return getTracer().withActiveSpan(
+      'DefaultUseCacheHandler.get',
+      async (span): ReturnType<CacheHandler['get']> => {
+        getLogger().withFields({ cacheKey }).debug(`[NetlifyDefaultUseCacheHandler] get`)
+        span.setAttributes({
+          cacheKey,
+        })
+
+        const pendingPromise = getPendingSets().get(cacheKey)
+        if (pendingPromise) {
+          await pendingPromise
+        }
+
+        const privateEntry = getLRUCache().get(cacheKey)
+        if (!privateEntry) {
+          getLogger()
+            .withFields({ cacheKey, status: 'MISS' })
+            .debug(`[NetlifyDefaultUseCacheHandler] get result`)
+          span.setAttributes({
+            cacheStatus: 'miss',
+          })
+          return undefined
+        }
+
+        const { entry } = privateEntry
+        const ttl = (entry.timestamp + entry.revalidate * 1000 - Date.now()) / 1000
+        if (ttl < 0) {
+          // In-memory caches should expire after revalidate time because it is
+          // unlikely that a new entry will be able to be used before it is dropped
+          // from the cache.
+          getLogger()
+            .withFields({ cacheKey, ttl, status: 'STALE' })
+            .debug(`[NetlifyDefaultUseCacheHandler] get result`)
+          span.setAttributes({
+            cacheStatus: 'expired, discarded',
+            ttl,
+          })
+          return undefined
+        }
+
+        if (await isAnyTagStale(entry.tags, entry.timestamp)) {
```

Comment:
Learning - If any tag is stale, then the entry is stale? So if I have a long chain of tags, if for example `_n_t_/layout` becomes stale then every page using that tag will be stale (since it's often the top level tag for an entry)?

Reply:
Yes, that's correct - so if you mark layout as stale - then all pages that use that layout are stale and should be regenerated on next opportunity